akka-typed/src/main/scala/akka/typed/cluster/Cluster.scala

```diff
@@ -0,0 +1,159 @@
+/**
+ * Copyright (C) 2009-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.typed.cluster
+
+import akka.actor.Address
+import akka.annotation.DoNotInherit
+import akka.cluster.ClusterEvent.{ ClusterDomainEvent, CurrentClusterState, InitialStateAsSnapshot, SubscriptionInitialStateMode }
+import akka.cluster._
+import akka.typed.internal.adapter.ActorSystemAdapter
+import akka.typed.{ ActorRef, ActorSystem, Extension, ExtensionId }
+
+import scala.collection.immutable
+import scala.reflect.ClassTag
+
+object Cluster extends ExtensionId[Cluster] {
+
+  sealed trait ClusterStateSubscription
+  // we can't have initial cluster as snapshot as that wouldn't be a ClusterDomainEvent?
+  object Subscribe {
+    def apply[T <: ClusterDomainEvent: ClassTag](
+      subscriber: ActorRef[T]) =
+      new Subscribe[T](subscriber, Set(implicitly[ClassTag[T]].runtimeClass))
+  }
+  final case class Subscribe[T <: ClusterDomainEvent](
+    subscriber: ActorRef[T],
+    // Not sure that/how we can enforce the clases, if we want multiple messages, but would be nice
+    // maybe just enforce using multiple messages in that case? or provide a small dsl "andAlso[OtherT]"
```

Comment:
if we only support `InitialStateAsEvents` I think there is only need for one class per `Subscribe`

Reply:
What are the downside of that, expensive if cluster is large?

Reply:
yes, that's the downside, but 100 messages or even 1000 messages is no big deal